File: app/src/main/java/com/borisruzanov/russianwives/mvp/model/repository/rating/RatingRepository.java
package com.borisruzanov.russianwives.mvp.model.repository.rating;

import android.support.annotation.NonNull;
import android.util.Log;

import com.borisruzanov.russianwives.utils.Consts;
import com.borisruzanov.russianwives.utils.RatingAchCallback;
import com.borisruzanov.russianwives.utils.RatingCallback;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.borisruzanov.russianwives.utils.FirebaseUtils.getUid;

public class RatingRepository {

    private DatabaseReference usersRef = FirebaseDatabase.getInstance().getReference().child(Consts.USERS_DB);

    public void addRating(int addPoint) {
        usersRef.child(getUid()).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                long rating;

                if (dataSnapshot.child(Consts.RATING).exists()) {
                    rating = dataSnapshot.child(Consts.RATING).getValue(Long.class);
                    long newRating = rating + addPoint;
                    dataSnapshot.child(Consts.RATING).getRef().setValue(newRating);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
            }
        });
    }

    public void addRating(int addPoint, String achievement) {
        usersRef.child(getUid()).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                long rating = 0;
                Map<String, String> achMap = new HashMap<>();
                List<String> achList = new ArrayList<>();

                if (dataSnapshot.child(Consts.RATING).exists()) {
                    rating = dataSnapshot.child(Consts.RATING).getValue(Long.class);
                    rating += addPoint;
                }

                if (dataSnapshot.child(Consts.ACHIEVEMENTS).exists()) {
                    achList.addAll((ArrayList<String>) dataSnapshot.child(Consts.ACHIEVEMENTS).getValue());
                    achList.addAll(achMap.values());
                    for (int i = 0; i < achMap.size(); i++) {
                        Log.d("RatingDebug", achMap.get(i));
                    }
                }
                achList.add(achievement);
                dataSnapshot.child(Consts.RATING).getRef().setValue(rating);
                dataSnapshot.child(Consts.ACHIEVEMENTS).getRef().setValue(achList);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
            }
        });
    }

    public void getRating(String uid, RatingCallback callback) {
        usersRef.child(uid).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.child(Consts.RATING).exists()) {
                    long rating = dataSnapshot.child(Consts.RATING).getValue(Long.class);
                    callback.setRating(rating);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
            }
        });
    }

    public void getRatingAch(RatingAchCallback callback) {
        usersRef.child(getUid()).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                long rating = 0;
                List<String> achievements = new ArrayList<>();
                if (dataSnapshot.child(Consts.RATING).exists()) {
                    rating = dataSnapshot.child(Consts.RATING).getValue(Long.class);
                }
                if (dataSnapshot.child(Consts.ACHIEVEMENTS).exists()) {
                    //Map<Object, String> achMap = (HashMap<Object, String>) dataSnapshot.getValue();
                    achievements.addAll((ArrayList<String>) dataSnapshot.child(Consts.ACHIEVEMENTS).getValue());
                }
                callback.setRatingAch(rating, achievements);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
            }
        });
    }

}

File: app/src/main/java/com/borisruzanov/russianwives/mvp/model/repository/rating/RatingRepository.kt
package com.borisruzanov.russianwives.mvp.model.repository.rating

import android.util.Log
import com.borisruzanov.russianwives.utils.BoolCallback

import com.borisruzanov.russianwives.utils.Consts
import com.borisruzanov.russianwives.utils.RatingAchievementsCallback
import com.borisruzanov.russianwives.utils.RatingCallback
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener

import java.util.ArrayList
import java.util.HashMap

import com.borisruzanov.russianwives.utils.FirebaseUtils.getUid
import com.google.android.gms.tasks.OnCompleteListener
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore

class RatingRepository {

    private val usersRef = FirebaseDatabase.getInstance().reference.child(Consts.USERS_DB)
    private val users = FirebaseFirestore.getInstance().collection(Consts.USERS_DB)

    fun isAchExist(uid: String, achievement: String, callback: BoolCallback) {
        usersRef.child(uid).addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                var flag = false
                if (dataSnapshot.child(Consts.ACHIEVEMENTS).exists()) {
                    val achievementsMap = dataSnapshot.child(Consts.ACHIEVEMENTS).value as MutableMap<String, Any>
                    val achievements = achievementsMap.keys
                    if (achievements.contains(achievement)) {
                        flag = true
                    }
                }
                callback.setBool(flag)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    fun isAchievementExist(achievement: String, callback: BoolCallback) {
        usersRef.child(getUid()).addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                var flag = false
                if (dataSnapshot.child(Consts.ACHIEVEMENTS).exists()) {
                    val achievementsMap = dataSnapshot.child(Consts.ACHIEVEMENTS).value as MutableMap<String, Any>
                    val achievements = achievementsMap.keys
                    if (achievements.contains(achievement)) {
                        flag = true
                    }
                }
                callback.setBool(flag)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    fun addRating(addPoint: Double) = addRating(getUid(), addPoint)

    fun addRating(uid: String, addPoint: Double) {
        usersRef.child(uid).addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val rating: Long

                if (dataSnapshot.child(Consts.RATING).exists()) {
                    rating = dataSnapshot.child(Consts.RATING).getValue(Long::class.java)!!
                    val newRating = rating + addPoint
                    dataSnapshot.child(Consts.RATING).ref.setValue(newRating)
                    //checkForAchive(age, gender, photo)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    fun checkForAchieve(params: List<String>, achievement: String) {
        var notDefault = true
        users.document(getUid()).get().addOnCompleteListener { task ->
            run {
                val snapshot: DocumentSnapshot? = task.result
                for (value in params) {
                    Log.d("AchDebug", "Value is ${snapshot?.getString(value)} " +
                            "and default is ${snapshot?.getString(value) == Consts.DEFAULT}")
                    if (snapshot?.getString(value) == Consts.DEFAULT) {
                        notDefault = false
                        break
                    }
                }
                if (notDefault) {
                    addAchievement(achievement)
                }
            }
        }
    }

    fun addAchievement(achievement: String) {
        usersRef.child(getUid()).addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                var achMap: MutableMap<String, Any> = HashMap()

                if (dataSnapshot.child(Consts.ACHIEVEMENTS).exists()) {
                    achMap = dataSnapshot.child(Consts.ACHIEVEMENTS).value as MutableMap<String, Any>
                }
                achMap[achievement] = true
                dataSnapshot.child(Consts.ACHIEVEMENTS).ref.setValue(achMap)
                addAchieveToFs(achievement)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    private fun addAchieveToFs(achieveName: String) {
        val achMap = HashMap<String, Any>()
        achMap[achieveName] = "true"
        users.document(getUid()).update(achMap)
    }

    fun getRating(uid: String, callback: RatingCallback) {
        usersRef.child(uid).addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.child(Consts.RATING).exists()) {
                    val rating = dataSnapshot.child(Consts.RATING).getValue(Long::class.java)!!
                    callback.setRating(rating)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    fun getRating(callback: RatingAchievementsCallback) {
        usersRef.child(getUid()).addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                var rating: Long = 0
                val achievements = ArrayList<String>()
                if (dataSnapshot.child(Consts.RATING).exists()) {
                    rating = dataSnapshot.child(Consts.RATING).getValue(Long::class.java)!!
                }
                if (dataSnapshot.child(Consts.ACHIEVEMENTS).exists()) {
                    val achMap = dataSnapshot.child(Consts.ACHIEVEMENTS).value as HashMap<String, Any>?
                    achievements.addAll(achMap!!.keys)
                }
                callback.setRatingAchievements(rating, achievements)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

}
